### PR TITLE
Add sideinputs to the RunInference Transform

### PR DESCRIPTION
**Pull request by AnandInguva**
Part 1 of https://github.com/apache/beam/issues/24042

* RunInference transform will accept side input emitting ModelMetadata.
* The Side input must emit one value at a time. It should be compatible with AsSingleton view
* All the metrics are tagged with unique id so that it would be easy to differentiate the metrics calculated by different models.
* Add a RunInference pipeline visitor to check if the side inputs are enabled only during --streaming mode and to. check if the side input has windows other than GlobalWindows with non default trigger.

This method follows slowly changing side input patterns as described in https://beam.apache.org/documentation/patterns/side-inputs/

TODO: 
Part 2: WatchFileTransform and it's use case. It will come in the following PRs.
Part 3: DLQ for error outputs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by AnandInguva** [comment]
Run Python 3.8 PostCommit

**Comment by AnandInguva** [comment]
R : @damccorm

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py]: This is causing postcommits to fail, should it be `model_metadata_pcoll`?

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base.py:75]: Why do we need this?

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base.py:86]:
```suggestion
PredictionResult.model_id.__doc__ = """Model ID used to run the prediction. Only populated when the model may be updated during the life of the pipeline."""
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base.py:89]: This is out of scope for this PR (and this work generally), but an interesting extension here would be to take a model class (or additional model metadata) for Model Handlers that require a class for loading the model. That way, you could start with a model of one type and switch to a model of different type during execution

Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:89]: I don't think its nearly as helpful as just retraining your model with new data (which this supports), but could be something we get asked about eventually. This approach should support it fine, so no action required here.

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base.py:540]: Do we actually need to check `_enable_side_input_loading` here? Won't this always return the same value whether we have ` and self._enable_side_input_loading`?

Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:540]: Or, alternatively can we ~just~ check `_enable_side_input_loading`?

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:86]: Ah, I see - you're actually always returning this. So maybe instead add something like `Only populated by model handlers that support updating the model during the life of the pipeline`?

File: sdks/python/apache_beam/ml/inference/pytorch_inference.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/pytorch_inference.py:48]: Nit: What's going on with the spacing here? Can we normalize this a bit?

File: sdks/python/apache_beam/ml/inference/pytorch_inference.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/pytorch_inference.py:117]: Thanks for moving this into utils

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (damccorm) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:238]:
```suggestion
From Beam 2.46.0, the RunInference PTransform will accept a side input of `ModelMetadata`, which is a `NamedTuple` containing the `model_id` and `model_name`.
```

2.45 has been cut already unfortunately.

Also, could we split website changes into a separate PR. Ideally we would merge those once the feature becomes available to avoid confusing people.

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py:393]: Note to self: I ran out of time to review this test, but don't want to block the rest of the feedback on this. I'll take another look at this when I have time.

**Review by damccorm — COMMENTED** [review]
Had some feedback, but on the whole this looks really good to me. Thanks Anand!

File: sdks/python/apache_beam/ml/inference/base.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base.py:540]: Yes, I want to use if `self._enable_side_input_loading` but mypy type checker is throwing errors since it interprets `si_model_metadata` as `Optional[ModelMetadata]`. So If I include `si_model_metadata`, the mypy errors goes off. 

We can just to if si_model_metadata as well. Let me see.

**Comment by AnandInguva** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/ml/inference/pytorch_inference.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/pytorch_inference.py]: image

The formatter is not letting me change it.

**Comment by AnandInguva** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/ml/inference/pytorch_inference.py
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/pytorch_inference.py:48]: Weird...

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:86]: SGTM

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base.py:75]: Thanks

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (damccorm) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:203]:
```suggestion
The `PredictionResult` is a `NamedTuple` object that contains both the input and the inferences, named  `example` and  `inference`, respectively. When keys are passed with the input data to the RunInference transform, the output `PCollection` returns a `Tuple[str, PredictionResult]`, which is the key and the `PredictionResult` object. Your pipeline interacts with a `PredictionResult` object in steps after the RunInference transform.
```

Looks like you missed this when removing website updates

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py:81]: Right now, this test doesn't actually show that this load_model method is getting invoked since this is basically a no-op after the first go-round. Could we rewrite this test so that FakeModel (or different fake model class) takes in the updated model id here and uses that to modify what it returns for its prediction?

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py:405]: `no output due to no SI` - This does still get an output, though, right? Since the side input still fires within this window

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py:425]: Should we be applying windowing on this PeriodicImpulse? In general our guidance is going to be to use the global window, right?

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py:394]: Should we be forcing this to be a streaming pipeline? Without perfect windowing, this won't work on Batch (and it doesn't make much sense), right?

Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base_test.py:394]: Relatedly, one thing that I think this PR is missing is validation that we are only running this in streaming mode. We'll probably need something similar to the `ExternalTransformFinder` to find RunInference with model updates at pipeline compile time - https://github.com/apache/beam/blob/3bf17574c6f18942d1015d5a837af23ae83c4d1a/sdks/python/apache_beam/pipeline.py#L1049 - or we'll need a way to read pipeline options during construction of this transform itself.

Its ok to add that in a separate PR, but we definitely need to do it (and document it)

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base_test.py:425]: (If so, this is a restriction we should probably enforce during pipeline construction)

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base_test.py:394]: I added `RunInferenceSideInputFinder`, which checks for `_enable_side_input_loading` bool. 

I think it is the simple way to go like you suggested. Thanks. 

if `_enable_side_input_loading` is enables and streaming flag not specified, we fail the pipeline during construction phase.

File: sdks/python/apache_beam/pipeline.py
Comment (damccorm) [inline sdks/python/apache_beam/pipeline.py:535]:
```suggestion
          "in streaming mode. To run in streaming mode, add the --streaming pipeline option.")
```

(may need reformatted to pass precommits)

File: sdks/python/apache_beam/pipeline.py
Comment (damccorm) [inline sdks/python/apache_beam/pipeline.py:533]: Could we please add a test to validate this gets correctly thrown?

**Comment by AnandInguva** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/ml/inference/base.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base.py]: Changed it.

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base_test.py:77]: Added 2 models, addition and sub but in a different IT test.

File: sdks/python/apache_beam/pipeline.py
Comment (AnandInguva) [reply sdks/python/apache_beam/pipeline.py:540]: I am good with error too.

**Review by AnandInguva — COMMENTED** [review]
PTAL @damccorm

File: sdks/python/apache_beam/examples/inference/run_inference_side_inputs.py
Comment (damccorm) [inline sdks/python/apache_beam/examples/inference/run_inference_side_inputs.py:125]: Nit: Please add an if/assert for the default model

File: sdks/python/apache_beam/pipeline.py
Comment (damccorm) [inline sdks/python/apache_beam/pipeline.py:549]:
```suggestion
          "GlobalWindows with a default trigger, non-global windowing, or no Windowing, which "
```

(probably needs formatting + corresponding test changes)

File: sdks/python/apache_beam/pipeline.py
Comment (damccorm) [reply sdks/python/apache_beam/pipeline.py:540]: I think I like the error that you have

**Comment by damccorm** [comment]
Run Python 3.8 PostCommit

**Comment by damccorm** [comment]
Run PythonDocker PreCommit

**Comment by damccorm** [comment]
Run Portable_Python PreCommit

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base_test.py]: LGTM

**Review by damccorm — COMMENTED** [review]
This mostly looks good to me, thanks! Just had a few small comments

**Comment by AnandInguva** [comment]
Run PythonDocker PreCommit

**Comment by AnandInguva** [comment]
Run Portable_Python PreCommit

**Comment by damccorm** [comment]
Precommit issues should be fixed by https://github.com/apache/beam/pull/25234#issuecomment-1410757836

**Comment by AnandInguva** [comment]
Run PythonDocker PreCommit

**Comment by AnandInguva** [comment]
Run Portable_Python PreCommit

**Comment by AnandInguva** [comment]
Run Python Dataflow V2 ValidatesRunner

**Comment by AnandInguva** [comment]
@damccorm I added the test to be in the validates DataflowRunner suite since I see that it is the test I see where all the side streaming tests are running.

Also, re the errors, i made a commit which I will push once the ValidateRunner test is done(Don't want to cancel in mid way).

PTAL

File: sdks/python/apache_beam/pipeline.py
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py:541]: Why?

File: sdks/python/apache_beam/pipeline.py
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py:528]: We should not be adding special support in pipeline.py for specific transforms. (E.g. this code will not get run when called via a cross-language transform.) Instead, we should be doing these checks at pipeline construction time (which will result in better, more local errors for the user too).

**Review by robertwb — CHANGES_REQUESTED** [review]

File: sdks/python/apache_beam/pipeline.py
Comment (AnandInguva) [reply sdks/python/apache_beam/pipeline.py:541]: For this side input, the target use cases would reading from unbounded sources or using PeriodicImpulse(which doesn't work as expected in batch mode). 

So by default, the model update side input is only restricted for streaming.

Comment (damccorm) [reply sdks/python/apache_beam/pipeline.py:541]: Specifically, because batch mode can completely process precondition stages before processing dependent stages, the meaning of pipeline updates breaks down (what does updating to the latest version of the model mean when all inferences are done at the "same time"), and if you do have a streaming source (which our built ins will be), the pipeline can hang forever.

File: sdks/python/apache_beam/pipeline.py
Comment (AnandInguva) [reply sdks/python/apache_beam/pipeline.py]: I moved away from the `pipeline.py` and added the logic during pipeline construction time. Thanks for the tip.

File: sdks/python/apache_beam/pipeline.py
Comment (robertwb) [reply sdks/python/apache_beam/pipeline.py]: In batch, the "latest" version of the model is likely to be the only version of the model, but that doesn't mean it can't be computed lazily. This also avoids having to have two separate codepaths (known at construction time can be "upgraded" to Create + Side Input). 

If you do have a streaming source, the pipeline should automatically run in streaming mode. (Also, IMHO, if just because a mode is more useful in one mode than the other doesn't mean we should prohibit it in the other unless it's actually detrimental.)

Comment (robertwb) [reply sdks/python/apache_beam/pipeline.py]: > > If you do have a streaming source, the pipeline should automatically run in streaming mode.
> 
> AFAIK, this isn't actually true in Python today.

If that's the case, we should definitely fix this :). 

> > if just because a mode is more useful in one mode than the other doesn't mean we should prohibit it in the other unless it's actually detrimental.
> 
> I think its detrimental in the batch use case because understanding the behavior requires understanding runner internals (in this case, how Dataflow/other runners handle stages), and the behavior is very confusing if you don't understand those internals. This is antithetical to one of the purposes of RunInference: taking a hard beam thing and making it easy.

Could you clarify a bit more why understanding runner internals is required for understanding what happens in the batch case (or, in other words, what confusing behavior the user would run into)? I'm not proposing we do away with the "easy" mode when the model is known at compile time (for batch or streaming), rather that we allow its computation to be deferred in both modes if this is explicitly what the user asks for. 

> I want RunInference users to be able to think about Beam as little as possible when building their pipelines, even if it comes at the cost of some of the purity of the unified model.

Yes, for sure. 

> One possible reconciliation here would be to emit a warning in batch mode with an informative error message, or allow users to intentionally enable running in batch mode with a parameter (though neither of those seem like awesome experiences, and I'd still personally favor explicitly disallowing this).

Comment (damccorm) [reply sdks/python/apache_beam/pipeline.py]: >>> If you do have a streaming source, the pipeline should automatically run in streaming mode.

>> AFAIK, this isn't actually true in Python today.

> If that's the case, we should definitely fix this :).

I tend to agree, but IIRC there were concerns about doing this (@tvalentyn thoughts?). It may have just been a matter of us not getting around to doing it yet. I would at least be worried about our existing IOs behaving correctly if we were to change this behavior, I'm not sure how crisp we've been here and I know of at least one (PeriodicImpulse/Sequence) that wasn't correctly marking itself unbounded. That's all solvable though, and probably isn't made worse by the change regardless.

> Could you clarify a bit more why understanding runner internals is required for understanding what happens in the batch case (or, in other words, what confusing behavior the user would run into)? I'm not proposing we do away with the "easy" mode when the model is known at compile time (for batch or streaming), rather that we allow its computation to be deferred in both modes if this is explicitly what the user asks for.

If the following things happened:

1) Pipeline starts with model A
2) 10,000 records arrive for inference + I do some preprocessing
3) Model updates to model B
4) 10,000 records arrive
5) Model updates to model C
6) 1 more records arrives

If I'm in batch mode, all 20,001 records get processed by model C (I actually think this depends on the runner, correct me if I'm wrong). If I'm in streaming mode, model A and model B each process 10,000 records, model C processes 1 record. The streaming case is by far the more useful one, and almost definitely the use case that most people are after with this kind of feature, but the discrepancy here is odd if you don't know how the model works.

With all that said, I can see cases where the batch use case would be nice (e.g. it allows daily pipelines without updating your model manually) - I think the framing of "deferred computation" vs "live model updates" helps there, and I could get behind allowing this. Though, the idea of a single unchanged transform for batch and streaming pipelines still seems fairly unrealistic to me because the boundedness of the side input source would need to be changed to match the boundedness of the main source in most cases to be useful.

My biggest sticking point is still that AFAIK unbounded sources don't automatically put pipelines in streaming mode, which leads to unexplained pipeline hangs in batch mode for this use case. Whether we change that or not, my vote would be to not block this PR on that change.

Comment (robertwb) [reply sdks/python/apache_beam/pipeline.py]: > > > > If you do have a streaming source, the pipeline should automatically run in streaming mode.
> 
> > > AFAIK, this isn't actually true in Python today.
> 
> > If that's the case, we should definitely fix this :).
> 
> I tend to agree, but IIRC there were concerns about doing this (@tvalentyn thoughts?). It may have just been a matter of us not getting around to doing it yet. I would at least be worried about our existing IOs behaving correctly if we were to change this behavior, I'm not sure how crisp we've been here and I know of at least one (PeriodicImpulse/Sequence) that wasn't correctly marking itself unbounded. That's all solvable though, and probably isn't made worse by the change regardless.

Filed https://github.com/apache/beam/issues/25264 ; we can continue discussion there (including if there's any downsides to this change, though I don't see any). If there is undesirable behavior, like hanging, when unbounded sources are used, this is entirely orthogonal to RunInferrence. 

> > Could you clarify a bit more why understanding runner internals is required for understanding what happens in the batch case (or, in other words, what confusing behavior the user would run into)? I'm not proposing we do away with the "easy" mode when the model is known at compile time (for batch or streaming), rather that we allow its computation to be deferred in both modes if this is explicitly what the user asks for.
> 
> If the following things happened:
> 
> 1. Pipeline starts with model A
> 2. 10,000 records arrive for inference + I do some preprocessing
> 3. Model updates to model B
> 4. 10,000 records arrive
> 5. Model updates to model C
> 6. 1 more records arrives
> 
> If I'm in batch mode, all 20,001 records get processed by model C (I actually think this depends on the runner, correct me if I'm wrong). If I'm in streaming mode, model A and model B each process 10,000 records, model C processes 1 record. The streaming case is by far the more useful one, and almost definitely the use case that most people are after with this kind of feature, but the discrepancy here is odd if you don't know how the model works.

One can't have this scenario for batch. In batch, one has 20,001 records and three models. None of the models (or records) are "first" and if anything there is ambiguity which model should be used. (Actually, with AsSingleton, an error will generally be raised if there's more than one.)

Now if the model (and records) are appropriately timestamped (this works in batch), they'll all get the "right" model. Bonus: this makes testing a lot easier. 

> With all that said, I can see cases where the batch use case would be nice (e.g. it allows daily pipelines without updating your model manually) - I think the framing of "deferred computation" vs "live model updates" helps there, and I could get behind allowing this. Though, the idea of a single unchanged transform for batch and streaming pipelines still seems fairly unrealistic to me because the boundedness of the side input source would need to be changed to match the boundedness of the main source in most cases to be useful.

A bounded side input with an unbounded main input is easy to understand, and the other way around can work too (especially if one is using timestamped data) though a bit odd. But we shouldn't fall into the trap that RunInferrence is always going to be at the top level (i.e. at the same level of the Reads)--it may be used in an even higher-level composite for example. 

> My biggest sticking point is still that AFAIK unbounded sources don't automatically put pipelines in streaming mode, which leads to unexplained pipeline hangs in batch mode for this use case. Whether we change that or not, my vote would be to not block this PR on that change.

If using an unbounded side input causes hangs, yes, let's warn on that (ideally as high a level as makes sense, it's not a RunInference-specific issue) until this is fixed. But I'd rather not prohibit side inputs altogether.

Comment (damccorm) [reply sdks/python/apache_beam/pipeline.py]: > Filed https://github.com/apache/beam/issues/25264 ;

Thanks

> None of the models (or records) are "first" and if anything there is ambiguity which model should be used.

I meant first by timestamp. For the model at least (if not the main input), I think this is a reasonable understanding of how the non-Beam person will see the world.

> Now if the model (and records) are appropriately timestamped (this works in batch), they'll all get the "right" model.

I didn't think this was the case, and it alleviates my concerns here.

> If using an unbounded side input causes hangs, yes, let's warn on that (ideally as high a level as makes sense, it's not a RunInference-specific issue) until this is fixed. But I'd rather not prohibit side inputs altogether.

I'm fine with this if Anand is.

Comment (AnandInguva) [reply sdks/python/apache_beam/pipeline.py]: I had an offline conversation with Danny with this and we both tend to agree on Robert's view.

I imposed the restriction because it would be easier to restrictive at first and then introduce new features(SI for batches) based on the feedback but now I think it is important to follow the beam unified model concept and remove the restriction on the side input.

Thanks

**Comment by AnandInguva** [comment]
@damccorm Added a couple of tests and removed the restrictions. 

PTAL.

File: sdks/python/apache_beam/examples/inference/run_inference_side_inputs.py
Comment (AnandInguva) [reply sdks/python/apache_beam/examples/inference/run_inference_side_inputs.py]: changed it. thanks

**Comment by AnandInguva** [comment]
Run Python Dataflow V2 ValidatesRunner

File: CHANGES.md
Comment (damccorm) [inline CHANGES.md:57]:
```suggestion
* RunInference PTransform will accept model paths as SideInputs in Python SDK. ([#24042](https://github.com/apache/beam/issues/24042))
```

**Review by damccorm — APPROVED** [review]

File: CHANGES.md
Comment (AnandInguva) [reply CHANGES.md]: Ah yes, thanks for catching it.

**Comment by damccorm** [comment]
I'll merge once checks pass

**Comment by damccorm** [comment]
Remaining checks all passed, they just didn't status for some reason

**Pull request merged**
